lib/OnyxCache.js

```diff
@@ -0,0 +1,150 @@
+import _ from 'underscore';
+import lodashMerge from 'lodash/merge';
+
+
+const isDefined = _.negate(_.isUndefined);
+
+/**
+ * In memory cache providing data by reference
+ * Encapsulates Onyx cache related functionality
+ */
+class OnyxCache {
+    constructor() {
+        /**
+         * @private
+         * Cache of all the storage keys available in persistent storage
+         * @type {Set<string>}
+         */
+        this.storageKeys = new Set();
+
+        /**
+         * @private
+         * A map of cached values
+         * @type {Record<string, *>}
+         */
+        this.storageMap = {};
+
+        /**
+         * @private
+         * Captured pending tasks for already running storage methods
+         * @type {Record<string, Promise>}
+         */
+        this.pendingPromises = {};
+
+        // bind all methods to prevent problems with `this`
+        _.bindAll(
+            this,
+            'getAllKeys', 'getValue', 'hasCacheForKey', 'addKey', 'set', 'remove', 'merge',
+            'hasPendingTask', 'getTaskPromise', 'captureTask',
+        );
+    }
+
+    /**
+     * Get all the storage keys
+     * @returns {string[]}
+     */
+    getAllKeys() {
+        return [...this.storageKeys];
+    }
+
+    /**
+     * Get a cached value from storage
+     * @param {string} key
+     * @returns {*}
+     */
+    getValue(key) {
+        return this.storageMap[key];
+    }
+
+    /**
+     * Check whether cache has data for the given key
+     * @param {string} key
+     * @returns {boolean}
+     */
+    hasCacheForKey(key) {
+        return isDefined(this.storageMap[key]);
+    }
+
+    /**
+     * Saves a key in the storage keys list
+     * Serves to keep the result of `getAllKeys` up to date
+     * @param {string} key
+     */
+    addKey(key) {
+        this.storageKeys.add(key);
+    }
+
+    /**
+     * Set's a key value in cache
+     * Adds the keys to the storage keys list as well
+     * Providing value is optional. Skipping value would just capture the key so that
+     * it's available in `getAllKeys`
+     * @param {string} key
+     * @param {*} value
+     * @returns {*} value - returns the cache value
+     */
+    set(key, value) {
```

Comment:
As per my other comment this is not intended to be optional, but a leftover documentation

You can still pass `undefined` or `null` but the handling is different

`undefined` would remove a value from cache - `set('myKey', undefined)`. E.g. if some iterative logic needs to update or remove values (like `Onyx.multiSet`). Otherwise prefer `remove('myKey')` 

`null` would set the cached value as `null` - it's a valid value and `hasCacheForKey` would return `true` in such case. For example `AsyncStorage.getItem` can return `null` if there's no data for a given `key` - in that case we still want to capture the result and prevent other callers from calling `AsyncStorage.getItem` again. We'll just tell them the result is `null` right away.